src/org/pentaho/platform/dataaccess/datasource/api/DatasourceService.java

```diff
@@ -25,41 +25,47 @@
 import org.pentaho.metadata.model.LogicalModel;
 import org.pentaho.metadata.repository.IMetadataDomainRepository;
 import org.pentaho.platform.api.engine.IAuthorizationPolicy;
-import org.pentaho.platform.api.engine.ISystemConfig;
 import org.pentaho.platform.api.engine.PentahoAccessControlException;
-import org.pentaho.platform.api.repository2.unified.RepositoryFile;
-import org.pentaho.platform.api.repository2.unified.RepositoryFileAcl;
+import org.pentaho.platform.api.repository2.unified.IUnifiedRepository;
+import org.pentaho.platform.dataaccess.datasource.utils.DataAccessPermissionUtil;
 import org.pentaho.platform.engine.core.system.PentahoSessionHolder;
 import org.pentaho.platform.engine.core.system.PentahoSystem;
 import org.pentaho.platform.plugin.action.mondrian.catalog.IMondrianCatalogService;
-import org.pentaho.platform.repository2.unified.jcr.IAclNodeHelper;
-import org.pentaho.platform.repository2.unified.jcr.JcrAclNodeHelper;
+import org.pentaho.platform.plugin.services.importexport.legacy.MondrianCatalogRepositoryHelper;
+import org.pentaho.platform.plugin.services.metadata.PentahoMetadataDomainRepository;
 import org.pentaho.platform.repository2.unified.webservices.RepositoryFileAclAdapter;
-import org.pentaho.platform.repository2.unified.webservices.RepositoryFileAclDto;
 import org.pentaho.platform.security.policy.rolebased.actions.AdministerSecurityAction;
 import org.pentaho.platform.security.policy.rolebased.actions.PublishAction;
 import org.pentaho.platform.security.policy.rolebased.actions.RepositoryCreateAction;
 import org.pentaho.platform.security.policy.rolebased.actions.RepositoryReadAction;
-import org.pentaho.platform.web.http.api.resources.services.FileService;
 
 public class DatasourceService {
-
+  // for usage of IAclAwarePentahoMetadataDomainRepositoryImporter
   protected IMetadataDomainRepository metadataDomainRepository;
+  // for flushing real metadata cache
+  private IMetadataDomainRepository metadataDomainRepositoryReal;
   protected IMondrianCatalogService mondrianCatalogService;
-  protected FileService fileService;
-  protected IAclNodeHelper aclHelper;
   protected RepositoryFileAclAdapter repositoryFileAclAdapter;
+  private MondrianCatalogRepositoryHelper mondrianCatalogRepositoryHelper;
 
   public DatasourceService() {
-    metadataDomainRepository = PentahoSystem.get( IMetadataDomainRepository.class, PentahoSessionHolder.getSession() );
+    metadataDomainRepositoryReal = PentahoSystem.get( IMetadataDomainRepository.class, PentahoSessionHolder.getSession() );
```

Comment:
We very rarely, almost never, directly construct an Implementation like this. What happens if an OEM has their own MetadataDomainRepository which needs to be used? A better solution is query for the "real" and the "plain" beans

metadataDomainRepositoryReal  = PentahoSystem.get( IMetadataDomainRepository.class, "IMetadataDomainRepositoryImpl", PentahoSessionHolder.getSession() );

metadataDomainRepository = PentahoSystem.get( IMetadataDomainRepository.class, "IMetadataDomainRepository", PentahoSessionHolder.getSession() );